drivers/wireless/bluetooth/bt_rpmsghci.c

```diff
@@ -0,0 +1,535 @@
+                      sizeof(struct rpmsghci_data_s) -1 + len);
+
+errout:
+  return ret;
+}
+
+/****************************************************************************
+ * Name: bt_rpmsghci_ioctl
+ *
+ * Description:
+ *   Rpmsg-HCI ioctl operation
+ *
+ * Parameters:
+ *   btdev  - the bt instance
+ *   cmd    - the ioctl command
+ *   arg    - the ioctl arguments
+ *
+ * Returned Values:
+ *   OK on success; A negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int bt_rpmsghci_ioctl(FAR struct bt_driver_s *btdev, int cmd,
+                              unsigned long arg)
+{
+  /* TODO */
+
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Name: rpmsghci_default_handler
+ *
+ * Description:
+ *   Default rpmsg-HCI response handler, this function will be called to
+ *   process the return message of rpmsghci_open(), rpmsghci_close(),
+ *   rpmsghci_send(), rpmsghci_ioctl().
+ *
+ * Parameters:
+ *   ept  - The rpmsg endpoint
+ *   data - The return message
+ *   len  - The return message length
+ *   src  - unknow
+ *   priv - unknow
+ *
+ * Returned Values:
+ *   Always OK
+ *
+ ****************************************************************************/
+
+static int rpmsghci_default_handler(FAR struct rpmsg_endpoint *ept,
+                                    FAR void *data, size_t len,
+                                    uint32_t src, FAR void *priv)
+{
+  FAR struct rpmsghci_header_s *header = data;
+  FAR struct rpmsghci_cookie_s *cookie = NULL;
+
+  if (header->cookie != 0)
+    {
+      cookie = (FAR struct rpmsghci_cookie_s *)(uintptr_t)header->cookie;
+      return rpmsg_post(ept, &cookie->sem);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rpmsghci_recv_handler
+ *
+ * Description:
+ *   Rpmsg-HCI receive handler.
+ *
+ * Parameters:
+ *   ept  - The rpmsg endpoint
+ *   data - The return message
+ *   len  - The return message length
+ *   src  - unknow
+ *   priv - unknow
+ *
+ * Returned Values:
+ *   OK on success; A negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int rpmsghci_recv_handler(FAR struct rpmsg_endpoint *ept,
+                                 FAR void *data, size_t len,
+                                 uint32_t src, FAR void *priv)
+{
+  FAR struct rpmsghci_s      *client = NULL;
+  FAR struct rpmsghci_data_s *msg    = data;
+  int                         ret    = OK;
+
+  DEBUGASSERT(priv);
+  client = priv;
+
+  wlinfo("rpmsghci_recv_handler %d\n", msg->type);
+
+  ret =  bt_netdev_receive(&client->btdev, msg->type, msg->data,
+                           len - sizeof(struct rpmsghci_data_s) + 1);
+  msg->header.result = ret;
+
+  return rpmsg_send(ept, msg, len);
```

Comment:
remove the response which is a hot path

Reply:
I have no idea how this will behave without confirmations

Reply:
In the case of an alarm callback, we don't have any return value, but `bt_netdev_receive` (`bt_hcicore.c::bt_receive`) can return an error code. With this approach, we lose this information, so I'm not sure about that change.